### PR TITLE
Dtolmaco/fix build response

### PR DESCRIPTION
**Pull request by dantol29**
## Now php_site works + buildResponse is more clean

### CODE REVIEW

**Review by 552020 — APPROVED** [review]
LGTM!

**Review by 552020 — CHANGES_REQUESTED** [review]
Great, thanks. I'll add an issue for me related the two errors where I'm tagged. Can we get just rid of 127.0.0.1 or localhost websites?

**Comment by dantol29** [comment]
## I noticed one problem:

I am able to connect to localhost:8081, however it is listening only to 8080. The problem is in the way Router finds a server block. We should also check port and ip, not only server_name @552020 @lmangall

File: var/127.0.0.1/index.html
[the remark this answers is not in the document]
Comment (dantol29) [reply var/127.0.0.1/index.html:1]: Ok

**Comment by 552020** [comment]
> ## I noticed one problem:
> I am able to connect to localhost:8081, however it is listening only to 8080. The problem is in the way Router finds a server block. We should also check port and ip, not only server_name @552020 @lmangall

We discussed this in detail with Leo. The request should be served only if the matching server block is listening to that ip:port. Otherwise is 404. I guess this was already implemented. But anyway it's not hard to implement it, we need just to compare the data on the server block with the data on the Connection object.

**Comment by dantol29** [comment]
> > ## I noticed one problem:
> > I am able to connect to localhost:8081, however it is listening only to 8080. The problem is in the way Router finds a server block. We should also check port and ip, not only server_name @552020 @lmangall
> 
> We discussed this in detail with Leo. The request should be served only if the matching server block is listening to that ip:port. Otherwise is 404. I guess this was already implemented. But anyway it's not hard to implement it, we need just to compare the data on the server block with the data on the Connection object.

The proper way would be to add ip:port to ServerBlock object

**Comment by 552020** [comment]
But the ServerBlock object has the Listen object(s) which have this info, no?

**Comment by dantol29** [comment]
## Now the way we find matching `ServerBlock` is correct

**Comment by dantol29** [comment]
I added default serve block logic, but I am still not sure about one thing.
```
if (serverName == "_" && defaultServerBlock == -1)
	defaultServerBlock = i;
```

If server block is not found, but one of them has `server_block "_"`, we serve default one. However, router is still trying to find folder with the name of the host like `html/www.random.com`. What should we do in this case? @552020

**Comment by lmangall** [comment]
If I understand well `_hasServerBlock` is used when we have a port and IP match, so the naming is misleading, perhaps `_isServerBlockMatched` or `_serverBlockFound`. 

Also I think the following code starting at server.cpp::243 
	```
if (response.getStatusCode() != 0)
	{
		Debug::log("Error response" + toString(response.getStatusCode()), Debug::NORMAL);
		// response.setErrorResponse(response.getStatusCode());
		router.handleServerBlockError(request, response, response.getStatusCode());
		conn.setHasDataToSend(true);
		return;
	}
```

can be deleted because it has been replaced by server.cpp::234 
 	```
if (!conn.getHasServerBlock())
		return (handleServerBlockError(conn, response));
```

Which also makes me wonder wether we need `_hasServerBlock`  altogether, since we use the error codes in the code already to catch errors. 
What do you think @552020 ? 
I think  `getHasServerBlock` doesn't particularly help with reading the function because we call `handleServerBlockError` (maybe this one should be renamed?) for a code different than 0.

@dantol29  If I remember well we can set a custom msg with the error code, so we could precise that it's a 404 because of a default to match the listen directive. However Stefano was rightfully against because this is giving potential hackers information we don't have to give.

**Comment by dantol29** [comment]
@lmangall We have to know if ServerBlock is found before we return an error, because ServerBlock may have custom errors.

**Comment by 552020** [comment]
> If I understand well `_hasServerBlock` is used when we have a port and IP match, so the naming is misleading, perhaps `_isServerBlockMatched` or `_serverBlockFound`.
> 
> Also I think the following code starting at server.cpp::243
> 
> ```
> if (response.getStatusCode() != 0)
> 	{
> 		Debug::log("Error response" + toString(response.getStatusCode()), Debug::NORMAL);
> 		// response.setErrorResponse(response.getStatusCode());
> 		router.handleServerBlockError(request, response, response.getStatusCode());
> 		conn.setHasDataToSend(true);
> 		return;
> 	}
> ```
> 
> can be deleted because it has been replaced by server.cpp::234
> 
> ```
> if (!conn.getHasServerBlock())
>    	return (handleServerBlockError(conn, response));
> ```
> 
> Which also makes me wonder wether we need `_hasServerBlock` altogether, since we use the error codes in the code already to catch errors. What do you think @552020 ? I think `getHasServerBlock` doesn't particularly help with reading the function because we call `handleServerBlockError` (maybe this one should be renamed?) for a code different than 0.
> 
> @dantol29 If I remember well we can set a custom msg with the error code, so we could precise that it's a 404 because of a default to match the listen directive. However Stefano was rightfully against because this is giving potential hackers information we don't have to give.

@lmangall I don't know if my answer to Daniil answers also this. I'm not familiar with the code of the router and even if we worked yesterady together with Daniil I already forgot the details of the implementation. But there is for sure a point where the Connection doesn't have a server block and then a point where the connection will have a server block and if not it will have one as the 'implicit' one. 

The connection will need to have a server block at least to send the proper error response.

**Comment by 552020** [comment]
And anyway @dantol29 to implement `default_server` and `_` are not part of the subject. The subject is asking only to make the default server if no matches for the host+ip:port are found the first server block with matching ip:port. So no need to implement the 'defaut_server` value for the listen directive and the '_' for the server_name. We can have though an implicit 'server block' serving the default error pages is no default server block in the config file is found.

**Comment by dantol29** [comment]
@lmangall Ok, I will delete `server_name "_"` logic. Then we just gonna server the first server block if none is found

**Comment by lmangall** [comment]
@552020 ok, thank you for the explanation.
@dantol29 That's not what I was talking about, this is new for me. But if you don't want to modify the logic to catch an error then I think you can remove the code I said in my previous message (server.cpp::243 and also the end of Connection::findServerBlock) because it is now useless. 

> The connection will need to have a server block at least to send the proper error response.

Does it ? So we won't use the error pages from our html directory in the root ?
Then `Server::handleServerBlockError` also need to be modified.

**Comment by dantol29** [comment]
> If I understand well `_hasServerBlock` is used when we have a port and IP match, so the naming is misleading, perhaps `_isServerBlockMatched` or `_serverBlockFound`.
> 
> Also I think the following code starting at server.cpp::243
> 
> ```
> if (response.getStatusCode() != 0)
> 	{
> 		Debug::log("Error response" + toString(response.getStatusCode()), Debug::NORMAL);
> 		// response.setErrorResponse(response.getStatusCode());
> 		router.handleServerBlockError(request, response, response.getStatusCode());
> 		conn.setHasDataToSend(true);
> 		return;
> 	}
> ```
> 
> can be deleted because it has been replaced by server.cpp::234
> 
> ```
> if (!conn.getHasServerBlock())
>    	return (handleServerBlockError(conn, response));
> ```
> 
> Which also makes me wonder wether we need `_hasServerBlock` altogether, since we use the error codes in the code already to catch errors. What do you think @552020 ? I think `getHasServerBlock` doesn't particularly help with reading the function because we call `handleServerBlockError` (maybe this one should be renamed?) for a code different than 0.
> 
> @dantol29 If I remember well we can set a custom msg with the error code, so we could precise that it's a 404 because of a default to match the listen directive. However Stefano was rightfully against because this is giving potential hackers information we don't have to give.

@lmangall @552020 I removed this piece of code. It is indeed unnecessary. Now if request server_name is not found we are looking for the first matching ip:port and take error_pages from there. Should be fine now

File: src/HTTPRequest.cpp
Comment (552020) [inline src/HTTPRequest.cpp:1]: But you know that this is bad style. Just saying. :-)

And it's not the only possibility!

**Review by 552020 — APPROVED** [review]
🥗

**Comment by 552020** [comment]
@lmangall Daniil kicked you out from the PR loop (but just temporarily) he said. Interns revolution!

**Comment by lmangall** [comment]
Hmmm I see I see, that was an urgent one to do. Well it's goo it got merged into main

**Pull request merged**